### PR TITLE
feature(#278): implement snackbar provider

### LINKED ISSUE #278: [Frontend] Investigate global error snackbar/toast
## Short description of the feature
Check to see if we can have a single maybe provider with a global state where errors can be sett to show better error messages on different pages.

[X] Relates to frontend
[ ] Relates to the backend
[ ] Requires modification of GraphQL
[ ] Requires modification of Database
[ ] Requires creating a new service

### PR DESCRIPTION
**Pull request by noahmay**
closes #278 

Also adds snackbar feedback to all mutations.

### CODE REVIEW

File: frontend/components/ArticleEditor/ArticleEditorItem/index.tsx [hunk 1]
```diff
@@ -32,17 +32,24 @@ export default function ArticleEditorItem({
   onImageChange,
   imageName,
 }: EditorProps) {
-  const classes = articleEditorItemStyles();
   const [fileName, setFileName] = React.useState('');
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/components/ArticleEditor/ArticleEditorItem/index.tsx:36]: `const { showMessage } = useSnackbar();` looks cleaner.

File: frontend/components/BookingForm/index.tsx [hunk 2]
```diff
@@ -26,11 +25,10 @@ export default function BookingForm({ onSubmit }: BookingFormProps) {
   const [startDateTime, setStartDateTime] = React.useState(DateTime.now());
   const [endDateTime, setEndDateTime] = React.useState(DateTime.now());
   const [bookables, setBookables] = React.useState<string[]>([]);
-  const [successOpen, setSuccessOpen] = React.useState(false);
-  const [errorOpen, setErrorOpen] = React.useState(false);
   const { user, loading: userLoading } = useContext(UserContext);
 
   const { data: bookableData, loading: bookableLoading } = useGetBookablesQuery();
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/components/BookingForm/index.tsx:31]: `const { showMessage } = useSnackbar();` looks cleaner.

File: frontend/components/BookingTable/bookingTableModifedStatusCell.tsx [hunk 3]
```diff
@@ -23,6 +24,7 @@ export default function bookingTableModifedStatusCell({
   ...rest
 }: BookingTableRowProps) {
   const { t } = useTranslation(['common, booking']);
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/components/BookingTable/bookingTableModifedStatusCell.tsx:27]: `const { showMessage } = useSnackbar();` looks cleaner.

File: frontend/public/locales/en/common.json [hunk 4]
```diff
@@ -45,5 +45,6 @@
   "add": "Add",
   "member": "Member",
   "role": "Role",
-  "YouDoNotHavePermissionToAccessThisPage": "You don't have permission to access this page"
-}
+  "YouDoNotHavePermissionToAccessThisPage": "You don't have permission to access this page",
+  "youDoNotHavePermissionToPreformThisAction": "You don't have permission to preform this action"
```
Comment (oliverlevay) [inline frontend/public/locales/en/common.json:49]: Change `preform` => `perform`.

I would prefer these to have shorter names like:
`no_permission_page`
`no_permission_action`

File: frontend/components/Calendar/EventEditor.tsx [hunk 5]
```diff
@@ -81,13 +80,36 @@ export default function EditEvent({ onSubmit, eventQuery }: BookingFormProps) {
   const english = selectedLanguage === 'en';
   const { user, loading: userLoading } = useContext(UserContext);
   const apiContext = useApiAccess();
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/components/Calendar/EventEditor.tsx:83]: const { showMessage } = useSnackbar(); looks cleaner.

File: frontend/components/FileBrowser/index.tsx [hunk 6]
```diff
@@ -52,15 +48,16 @@ export default function Browser({ bucket }: Props) {
   const [uploadFiles, setUploadFiles] = useState<File[]>([]);
 
   const { t, i18n } = useTranslation(['common', 'fileBrowser']);
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/components/FileBrowser/index.tsx:51]: const { showMessage } = useSnackbar(); looks cleaner.

File: frontend/components/Positions/CreateMandate.tsx [hunk 7]
```diff
@@ -23,9 +24,11 @@ const CreateMandate = ({
 }) => {
   const [startDate, setStartDate] = useState(defaultFromDate);
   const [endDate, setEndDate] = useState(defaultToDate);
-  const { t, i18n } = useTranslation(['common']);
+  const { t, i18n } = useTranslation(['common', 'committee']);
   const { refetchMandates } = useCurrentMandates();
   const [selectedMemberToAdd, setSelectedMemberToAdd] = useState<number>(null);
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/components/Positions/CreateMandate.tsx:30]: const { showMessage } = useSnackbar(); looks cleaner.

File: frontend/pages/members/[id]/edit.tsx [hunk 8]
```diff
@@ -32,8 +31,7 @@ export default function EditMemberPage() {
   const [classProgramme, setClassProgramme] = useState('');
   const [classYear, setClassYear] = useState('');
   const [picturePath, setPicturePath] = useState('');
-  const [successOpen, setSuccessOpen] = React.useState(false);
-  const [errorOpen, setErrorOpen] = React.useState(false);
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/pages/members/[id]/edit.tsx:34]: const { showMessage } = useSnackbar(); looks cleaner.

File: frontend/pages/news/article/[id]/edit.tsx [hunk 9]
```diff
@@ -33,6 +32,7 @@ export default function EditArticlePage() {
   });
 
   const { user, loading: userLoading } = useContext(UserContext);
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/pages/news/article/[id]/edit.tsx:35]: const { showMessage } = useSnackbar(); looks cleaner.

File: frontend/pages/news/article/create.tsx [hunk 10]
```diff
@@ -33,8 +32,8 @@ export default function CreateArticlePage() {
   const [header, setHeader] = useState({ sv: '', en: '' });
   const [imageFile, setImageFile] = useState<File | undefined>(undefined);
   const [imageName, setImageName] = useState('');
-  const [successOpen, setSuccessOpen] = useState(false);
-  const [errorOpen, setErrorOpen] = useState(false);
+  const snackbarContext = useSnackbar();
```
Comment (oliverlevay) [inline frontend/pages/news/article/create.tsx:35]: I'm going to stop commenting all of these, just do a search & replace

**Review by oliverlevay — CHANGES_REQUESTED** [review]
Just a few things I'd like you to change.

**Comment by oliverlevay** [comment]
I have changed too many things in this PR to be a legitimate reviewer.

File: frontend/components/ArticleEditor/ArticleEditorItem/index.tsx [hunk 11]
```diff
@@ -35,15 +36,23 @@ export default function ArticleEditorItem({
...
+  const { showMessage } = useSnackbar();
+  const { t } = useTranslation(['common', 'news']);
 
   const [getPresignedPutUrlMutation] = useGetPresignedPutUrlMutation({
     variables: {
       fileName,
     },
+    onError: (error) => {
+      console.error(error.message);
+      if (error.message.includes('You do not have permission')) {
```
Comment (emilwihlander) [inline frontend/components/ArticleEditor/ArticleEditorItem/index.tsx:48]: There must be a better way to do this, maybe through custom error codes?

File: frontend/components/BookingTable/bookingTableModifedStatusCell.tsx [hunk 12]
```diff
@@ -32,6 +34,17 @@ export default function BookingTableModifedStatusCell({
     variables: {
       id: bookingId,
     },
+    onCompleted: () => {
+      showMessage(t('booking:requestDenied'), 'success');
+    },
+    onError: (error) => {
+      console.error(error.message);
+      if (error.message.includes('You do not have permission')) {
+        showMessage(t('common:youDoNotHavePermissionToPreformThisAction'), 'error');
```
Comment (emilwihlander) [inline frontend/components/BookingTable/bookingTableModifedStatusCell.tsx:43]: Spelling mistake is still in there

File: frontend/components/Calendar/EventEditor.tsx [hunk 13]
```diff
@@ -1,4 +1,5 @@
-import React, { useContext, useEffect, useState } from 'react';
+/* eslint-disable @typescript-eslint/no-use-before-define */
```
Comment (emilwihlander) [inline frontend/components/Calendar/EventEditor.tsx:1]: Is this necessary?

File: frontend/providers/SnackbarProvider.tsx [hunk 14]
```diff
@@ -0,0 +1,73 @@
+import React, { useCallback, useState } from 'react';
+import Snackbar from '@mui/material/Snackbar';
+import MuiAlert from '@mui/material/Alert';
```
Comment (emilwihlander) [inline frontend/providers/SnackbarProvider.tsx:3]: We typically name them without "Mui"

**Review by emilwihlander — APPROVED** [review]
The spelling mistake (it might also be the wrong key, saw another key used in other places) must be fixed, otherwise it looks good!

File: frontend/components/ArticleEditor/ArticleEditorItem/index.tsx (hunk 11 above)
Comment (emilwihlander) [reply frontend/components/ArticleEditor/ArticleEditorItem/index.tsx:48]: Errors codes apparently exists by default: https://www.apollographql.com/docs/apollo-server/data/errors/

This looks cleaner imo:
```typescript
if (error.extensions.code === 'UNAUTHENTICATED') {
 ...
}
``` 

Might even be worth having helper functions:
```typescript
function unauthenticated(error: /*some type*/) {
  return error.extensions.code === 'UNAUTHENTICATED';
}
function badUserInput(error: /*some type*/) {
  return error.extensions.code === 'BAD_USER_INPUT';
}
...
if (unauthenticated(error)) {
 ...
}
```

**Review by oliverlevay — APPROVED** [review]

File: frontend/components/Calendar/EventEditor.tsx (hunk 13 above)
Comment (oliverlevay) [reply frontend/components/Calendar/EventEditor.tsx:1]: Well I'm  sure we can attempt to build it away, I gave it like 10 minutes and then I gave up. You are very welcome to give it a try!

File: frontend/components/ArticleEditor/ArticleEditorItem/index.tsx (hunk 11 above)
Comment (oliverlevay) [reply frontend/components/ArticleEditor/ArticleEditorItem/index.tsx]: this is the solution I ended up going for
```typescript
const handleApolloError = (
  apolloError: ApolloError,
  showMessage: (
    message: string,
    severity: 'success' | 'info' | 'warning' | 'error'
  ) => void,
  t: TFunction,
  genericErrorKey: string = 'common:error',
) => {
  const { graphQLErrors, message } = apolloError;
  console.error(message);
  graphQLErrors.forEach((error) => {
    switch (error.extensions.code) {
      case 'FORBIDDEN':
        showMessage(t('common:no_permission_action'), 'error');
        break;
      default:
        showMessage(t(genericErrorKey), 'error');
        break;
    }
  });
};
```

Comment (oliverlevay) [reply frontend/components/ArticleEditor/ArticleEditorItem/index.tsx]: in hindsight perhaps this function should just return a translation key and the consumer can do as they wish with it...

**Pull request merged**